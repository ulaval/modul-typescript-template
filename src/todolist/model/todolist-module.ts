import { Action, ModuleBase, Mutation } from '@ulaval/modul-vuex/dist/vuex-annotations';
import Vue from 'vue';
import { Store } from 'vuex';
import { FormMode, Todo, Todolist, TodolistFormState, TodolistState } from './todolist-model';

export interface TodolistRepository {
    loadTodolist(todolistId: string): Promise<Todolist>;
    searchTodos(todolistId: string, lastTodolistId?: string, pageSize?: number): Promise<Todo[]>;
    addTodo(todo: Todo): Promise<Todo>;
    updateTodo(todo: Todo): Promise<Todo>;
    deleteTodo(todo: Todo): Promise<Todo>;
}

interface TodolistModuleState {
    todos: { [k: string]: Todo | undefined };

    todolists: { [k: string]: TodolistState | undefined };
}

export class TodolistModule extends ModuleBase<TodolistModuleState> {
    constructor(private todolistRepository: TodolistRepository, store: Store<any>, moduleName: string = 'Todolist') {
        super(moduleName, {
            todolists: {},
            todos: {}
        }, store);
    }

    public getTodolistState(todolistId: string): TodolistState | undefined {
        return this.state.todolists[todolistId];
    }

    @Mutation()
    public onOpenNewForm(todolistState: TodolistState, mode: FormMode = 'new'): void {
        todolistState.todolistFormState = new TodolistFormState();
        todolistState.todolistFormState.open = true;
        todolistState.todolistFormState.mode = mode;
    }

    @Mutation()
    public onConfirmForm(state: TodolistState, todo: Todo): void {
        state.visibleTodos.push(todo);
        state.todolistFormState = new TodolistFormState();
    }

    @Mutation()
    public onCloseForm(state: TodolistState): void {
        state.todolistFormState = new TodolistFormState();
    }

    @Mutation()
    public onDeleteTodo(state: TodolistState, todo: Todo): void {
        const index: number = state.visibleTodos.findIndex((val: Todo) => val === todo);

        if (index >= 0) {
            Vue.delete(state.visibleTodos, index);
        }
    }

    @Action()
    public async loadTodolist(todolistId: string): Promise<void> {

        this.setLoadingTodoList(todolistId, true);

        try {
            const todolist: Todolist = await this.todolistRepository.loadTodolist(todolistId);
            const visibleTodos: Todo[] = await this.todolistRepository.searchTodos(todolistId);

            this.updateTodoListState(todolist, visibleTodos);
        } catch (e) {
            this.setLoadingTodoList(todolistId, false, e);
            throw e;
        }
    }

    @Action()
    public loadMore(todolistId: string): void {

    }

    @Action()
    public addTodo(todo: Todo): void {

    }

    @Action()
    public updateTodo(todo: Todo): void {

    }

    @Action()
    public deleteTodo(todo: Todo): void {

    }

    @Mutation()
    private setLoadingTodoList(todolistId: string, loading: boolean, error: any = undefined): void {
        let todolistState: TodolistState | undefined = this.state.todolists[todolistId];

        if (!todolistState) {
            todolistState = new TodolistState();
            Vue.set(this.state.todolists, todolistId, todolistState);
        }

        todolistState.setLoading(loading);

        if (error) {
            todolistState.loadingError = true;
        }
    }

    @Mutation()
    private updateTodoListState(todolist: Todolist, visibleTodos: Todo[]): void {
        if (!todolist.todolistId) {
            throw new Error('Invalid todolist');
        }

        let currentState: TodolistState | undefined = this.getTodolistState(todolist.todolistId);

        if (currentState) {
            currentState.todolist = todolist;
        } else {
            currentState = new TodolistState();
            currentState.todolist = todolist;
            this.state.todolists[todolist.todolistId] = currentState;
        }

        currentState.visibleTodos = visibleTodos;
        currentState.setLoading(false);
    }
}
