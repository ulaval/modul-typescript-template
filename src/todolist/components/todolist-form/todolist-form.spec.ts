import { TodolistFormState } from '@/todolist/model/todolist-model';
import { genSnapshotsTests, PropsData } from '../../../../jest/jest-utils';
import TodolistForm from './todolist-form';


const states: PropsData = {
    'new state': {
        state: {
            mode: 'new',
            open: false
        } as TodolistFormState
    },
    'edit state': {
        state: {
            mode: 'edit',
            open: false
        } as TodolistFormState
    }
};

describe('TodolistForm', () => {
    genSnapshotsTests(TodolistForm, states);
});
