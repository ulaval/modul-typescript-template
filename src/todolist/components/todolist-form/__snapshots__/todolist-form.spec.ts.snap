// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodolistForm edit state 1`] = `
<m-modal title="Edit">
  <section>
    Title: <m-textfield></m-textfield> <br>
    Description: <m-textfield></m-textfield>
    <m-button>Ok</m-button>
    <m-button>Cancel</m-button>
  </section>
</m-modal>
`;

exports[`TodolistForm new state 1`] = `
<m-modal title="New">
  <section>
    Title: <m-textfield></m-textfield> <br>
    Description: <m-textfield></m-textfield>
    <m-button>Ok</m-button>
    <m-button>Cancel</m-button>
  </section>
</m-modal>
`;
